lib/bolt/module_installer/puppetfile.rb

```diff
@@ -97,20 +97,43 @@ def assert_satisfies(specs)
           end
         end
 
-        return if unsatisfied_specs.empty?
-
+        #versionless_mods = @modules.any? do |mod|    
+        #  mod.version.nil? 
+        #end
+        versionless_mods = []
+        @modules.each {|mod|
+          if mod.version.nil?
+            versionless_mods << mod
+          end
+        }
+        
+        return if unsatisfied_specs.empty? && versionless_mods.empty?
         command = Bolt::Util.windows? ? 'Install-BoltModule -Force' : 'bolt module install --force'
+        
+        if !unsatisfied_specs.empty?
+          message = <<~MESSAGE.chomp
+            Puppetfile does not include modules that satisfy the following specifications:
+
+            #{unsatisfied_specs.map(&:to_hash).to_yaml.lines.drop(1).join.chomp}
+            
+            This Puppetfile might not be managed by Bolt. To forcibly overwrite the
+            Puppetfile, run '#{command}'.
+          MESSAGE
+
+          raise Bolt::Error.new(message, 'bolt/missing-module-specs')
+
+        else
+          message = <<~MESSAGE.chomp
+            Puppetfile includes Forge modules without a version requirement:
+            #{versionless_mods.map(&:to_hash).to_yaml.lines.drop(1).join.chomp}
```

Comment:
```suggestion
            #{versionless_mods.map(&:name).join.chomp}
```

`to_hash` will format the modules similar to how they'd be written in a `bolt-project.yaml` file, which is a bit misleading. Instead, we just want to list the module names that are missing the version requirement.